<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
امید به زندگی
</font>
</h1>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
مقدمه و صورت مسئله
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
به تمرین «امید به زندگی» خوش آمدید :)
    <br>
    هدف این تمرین این است که با مورد بررسی قرار دادن شاخص‌های بهداشتی، مالی و کشوری، امید به زندگی کشورها را پیش‌بینی کنیم.
    <br>
    امید به زندگی یا متوسط طول عمر (به انگلیسی: Life Expectancy) یک شاخص آماری است که نشان می‌دهد اعضای یک جامعه به‌طور میانگین چقدر عمر می‌کنند، یا به عبارت دیگر، انتظار می‌رود چقدر عمر کنند. اطلاعات بیشتر درباره‌ی امید به زندگی را می‌توانید در 
<a href="https://fa.wikipedia.org/wiki/%D8%A7%D9%85%DB%8C%D8%AF_%D8%A8%D9%87_%D8%B2%D9%86%D8%AF%DA%AF%DB%8C">ویکی‌پدیا</a> 
    مطالعه کنید.
    
</font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
وارد کردن کتابخانه‌های مورد نیاز
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    ابتدا کتابخانه‌های مورد نیازتان را وارد کنید.
</font>
</p>

In [110]:
import numpy as np
import pandas as pd 

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
معرفی مجموعه داده
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
مجموعه‌داده‌ی آموزشی مسئله دارای ۲۸۴۸ سطر و ۱۸ ستون است.
    ۱۷ ستون مربوط به متغیرهای مستقل و ستون آخر هم نشانگر متغیر وابسته مسئله است.
    <br>
    در جدول زیر توضیح هر ستون را مشاهده می‌کنید. هر سطر مربوط به اطلاعات یک کشور است که در یک سال مشخص رکورد شده است.
</font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>Country</code>|کشور مورد مطالعه|
|<code>Year</code>|سال|
|<code>Status</code>|وضعیت توسعه‌یافتگی کشور|
|<code>Population</code>|جمعیت کشور|
|<code>Hepatitis B</code>|درصد افراد نهایتاً یک ساله‌ای که در برابر هپاتیت ب ایمن شده‌اند|
|<code>Measles</code>|تعداد موارد مبتلا به سرخک گزارش شده در هر ۱۰۰۰ نفر|
|<code>Polio</code>|درصد افراد نهایتاً یک ساله‌ای که در برابر فلج اطفال ایمن شده‌اند|
|<code>Diphtheria</code>|درصد افراد نهایتاً یک ساله‌ای که در برابر دیفتری ایمن شده‌اند|
|<code>HIV/AIDS</code>|تعداد فوتی‌های ناشی از ایدز افراد نهایتاً ۴ ساله‌ای که زنده به دنیا آمده‌اند در هر ۱۰۰۰ نفر|
|<code>infant deaths</code>|تعداد فوتی‌های نوزادان در هر ۱۰۰۰ نفر|
|<code>under-five deaths</code>|تعداد فوتی‌های افراد نهایتاً ۵ ساله در هر ۱۰۰۰ نفر|
|<code>Total expenditure</code>|نسبت هزینه‌های درمانی-بهداشتی دولت به کل هزینه‌های دولت بر حسب درصد|
|<code>GDP</code>|تولید ناخالص ملی بر حسب دلار امریکا|
|<code>BMI</code>|میانگین شاخص توده بدنی کل جمعیت کشور|
|<code>thinness 1-19 years</code>|میزان شیوع لاغری میان افراد نهایتاً ۱۹ ساله بر حسب درصد|
|<code>Alcohol</code>|میزان لیتر مصرف الکل بین افراد بیشتر از ۱۵ سال|
|<code>Schooling</code>|تعداد سالی که افراد تحصیل می‌کنند|
|<code>Life expectancy</code>|امید به زندگی یا متوسط طول عمر|
    
</font>
</div>
</center>


<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
     مجموعه‌ی آزمون (تست) مانند مجموعه‌ی آموزش است با این تفاوت که ستون <code>Life expectancy</code> را ندارد.
</font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
خواندن مجموعه داده
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    در ابتدا نیاز است فایل‌های مجموعه‌داده را بخوانید. نمونه‌های آموزشی در فایل <code>train.csv</code> و نمونه‌های آزمون که باید مقدار متغیر هدف آن‌ها را پیش‌بینی کنید در فایل <code>test.csv</code> ذخیره شده‌اند. اگر لازم دانستید می‌توانید به دلخواه خود بخشی از مجموعه‌ی آموزش را به عنوان مجموعه‌ی اعتبارسنجی نیز جدا کنید.
</font>
</p>

In [111]:
train_data = pd.read_csv('../data/train.csv')
train_data

Country  Year      Status  Population  Hepatitis B  Measles  Polio  \
0     Afghanistan  2015  Developing  33736494.0         65.0     1154    6.0   
1     Afghanistan  2014  Developing    327582.0         62.0      492   58.0   
2     Afghanistan  2013  Developing  31731688.0         64.0      430   62.0   
3     Afghanistan  2012  Developing   3696958.0         67.0     2787   67.0   
4     Afghanistan  2011  Developing   2978599.0         68.0     3013   68.0   
...           ...   ...         ...         ...          ...      ...    ...   
2843     Zimbabwe  2004  Developing  12777511.0         68.0       31   67.0   
2844     Zimbabwe  2003  Developing  12633897.0          7.0      998    7.0   
2845     Zimbabwe  2002  Developing    125525.0         73.0      304   73.0   
2846     Zimbabwe  2001  Developing  12366165.0         76.0      529   76.0   
2847     Zimbabwe  2000  Developing  12222251.0         79.0     1483   78.0   

      Diphtheria  HIV/AIDS  infant deaths  under-five deaths  \
0           65.0       0.1             62                 83   
1           62.0       0.1             64                 86   
2           64.0       0.1             66                 89   
3           67.0       0.1             69                 93   
4           68.0       0.1             71                 97   
...          ...       ...            ...                ...   
2843        65.0      33.6             27                 42   
2844        68.0      36.7             26                 41   
2845        71.0      39.8             25                 40   
2846        75.0      42.1             25                 39   
2847        78.0      43.5             24                 39   

      Total expenditure         GDP   BMI  thinness  1-19 years  Alcohol  \
0                  8.16  584.259210  19.1                  17.2     0.01   
1                  8.18  612.696514  18.6                  17.5     0.01   
2                  8.13  631.744976  18.1                  17.7     0.01   
3                  8.52  669.959000  17.6                  17.9     0.01   
4                  7.87   63.537231  17.2                  18.2     0.01   
...                 ...         ...   ...                   ...      ...   
2843               7.13  454.366654  27.1                   9.4     4.36   
2844               6.52  453.351155  26.7                   9.8     4.06   
2845               6.53   57.348340  26.3                   1.2     4.43   
2846               6.16  548.587312  25.9                   1.6     1.72   
2847               7.10  547.358878  25.5                  11.0     1.68   

      Schooling  Life expectancy  
0          10.1             65.0  
1          10.0             59.9  
2           9.9             59.9  
3           9.8             59.5  
4           9.5             59.2  
...         ...              ...  
2843        9.2             44.3  
2844        9.5             44.5  
2845       10.0             44.8  
2846        9.8             45.3  
2847        9.8             46.0  

[2848 rows x 18 columns]

In [112]:
test_data = pd.read_csv('../data/test.csv')
test_data

Country  Year      Status  Population  Hepatitis B  Measles  \
0            Burundi  2015  Developing    119927.0         94.0        9   
1            Burundi  2014  Developing    989179.0         95.0        0   
2            Burundi  2013  Developing     96186.0         96.0        0   
3            Burundi  2012  Developing    931971.0         96.0       49   
4            Burundi  2011  Developing     94358.0         96.0      129   
..               ...   ...         ...         ...          ...      ...   
75  Papua New Guinea  2004  Developing   6161517.0          6.0     1385   
76  Papua New Guinea  2003  Developing     61724.0         67.0     3863   
77  Papua New Guinea  2002  Developing   5862316.0          6.0    17620   
78  Papua New Guinea  2001  Developing   5716152.0         59.0     4023   
79  Papua New Guinea  2000  Developing   5572222.0         57.0     7135   

    Polio  Diphtheria  HIV/AIDS  infant deaths  under-five deaths  \
0    94.0        94.0       0.7             21                 31   
1    95.0        95.0       0.7             22                 32   
2    96.0        96.0       1.0             22                 32   
3    96.0        96.0       1.2             22                 33   
4    95.0        96.0       1.5             22                 33   
..    ...         ...       ...            ...                ...   
75   59.0        62.0       1.5             11                 15   
76   62.0        68.0       1.4             11                 14   
77   54.0        61.0       1.4             11                 14   
78   52.0        55.0       1.3             11                 14   
79   51.0        59.0       1.1             11                 14   

    Total expenditure         GDP   BMI  thinness  1-19 years  Alcohol  \
0                 NaN   33.681223  18.7                   7.3      NaN   
1                7.54  312.748979  18.2                   7.4     0.01   
2                8.30  282.755525  17.6                   7.4     0.01   
3                8.21  265.285651  17.1                   7.5     0.01   
4                8.58   26.479973  16.6                   7.6     4.16   
..                ...         ...   ...                   ...      ...   
75               8.39  637.361621   4.4                   1.5     0.58   
76               6.79  588.358260  39.6                   1.5     0.59   
77               7.69  511.665773  38.9                   1.5     0.61   
78               6.68  539.415280  38.2                   1.5     0.68   
79               3.98  631.946853  37.5                   1.6     0.73   

    Schooling  
0        10.6  
1        10.6  
2        10.5  
3        10.3  
4         9.9  
..        ...  
75        7.6  
76        7.2  
77        6.9  
78        6.6  
79        6.2  

[80 rows x 17 columns]

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
پیش‌پردازش و مهندسی ویژگی
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    در این سوال شما می‌توانید از هر تکنیک پیش‌پردازش/مهندسی ویژگی که در فصل‌های گذشته آموختید، استفاده کنید.
    <br>
    تکنیک‌هایی که استفاده می‌کنید به شکل مستقیم مورد ارزیابی توسط سامانه داوری قرار <b>نمی‌گیرند.</b> بلکه همه آن‌ها در دقت مدل شما تاثیر خواهند گذاشت؛ بنابراین هر چه پیش‌پردازش/مهندسی ویژگی بهتری انجام دهید تا دقت مدل بهبود پیدا کند، امتیاز بیشتری از این سوال کسب خواهید کرد.
</font>
</p>

In [113]:
# calc nan value

train_data.isna().sum()

Country                   0
Year                      0
Status                    0
Population              644
Hepatitis B             542
Measles                   0
Polio                    19
Diphtheria               19
HIV/AIDS                  0
infant deaths             0
under-five deaths         0
Total expenditure       221
GDP                     442
BMI                      32
thinness  1-19 years     32
Alcohol                 188
Schooling               160
Life expectancy           0
dtype: int64

In [114]:
test_data.isna().sum()

Country                  0
Year                     0
Status                   0
Population               0
Hepatitis B             11
Measles                  0
Polio                    0
Diphtheria               0
HIV/AIDS                 0
infant deaths            0
under-five deaths        0
Total expenditure        5
GDP                      1
BMI                      0
thinness  1-19 years     0
Alcohol                  5
Schooling                0
dtype: int64

In [116]:
saved1 = lambda group: group.fillna(group.median())

train_data = train_data.groupby('Country').apply(lambda group: group.fillna(group.median()))

train_data = train_data.fillna(train_data.median())
saved2 = train_data.median()


train_data.isna().sum()


C:\Users\Danial\AppData\Local\Temp\ipykernel_12512\3158041821.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  train_data = train_data.groupby('Country').apply(lambda group: group.fillna(group.median()))
C:\Users\Danial\AppData\Local\Temp\ipykernel_12512\3158041821.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  train_data = train_data.groupby('Country').apply(lambda group: group.fillna(group.median()))
C:\Users\Danial\AppData\Local\Temp\ipykernel_12512\3158041821.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid co

Country                 0
Year                    0
Status                  0
Population              0
Hepatitis B             0
Measles                 0
Polio                   0
Diphtheria              0
HIV/AIDS                0
infant deaths           0
under-five deaths       0
Total expenditure       0
GDP                     0
BMI                     0
thinness  1-19 years    0
Alcohol                 0
Schooling               0
Life expectancy         0
dtype: int64

In [118]:

i = 0
for j in test_data.values:
    if not(test_data.iloc[i][0] in train_data.reset_index(drop=True).Country.values):
        break
    print(test_data.iloc[i][0])
    test_data.iloc[i] = test_data.iloc[i].fillna(train_data.loc[test_data.iloc[i][0]].median())
    i = i+1


# test_data = test_data.groupby('Country').apply(saved1)

test_data = test_data.fillna(saved2)

test_data.isna().sum()

Country                 0
Year                    0
Status                  0
Population              0
Hepatitis B             0
Measles                 0
Polio                   0
Diphtheria              0
HIV/AIDS                0
infant deaths           0
under-five deaths       0
Total expenditure       0
GDP                     0
BMI                     0
thinness  1-19 years    0
Alcohol                 0
Schooling               0
dtype: int64

In [119]:
train_data.Country.value_counts()

dict = {}
x = 0;

def country_mapping(name):

    if not(name in dict):
        global  x
        dict[name] = x
        x = x + 1



train_data.Country.apply(country_mapping)
train_data.Country.replace(dict,inplace=True)


test_data.Country.apply(country_mapping)
test_data.Country.replace(dict,inplace=True)



Status_mapping = {
    'Developing': 0,
    'Developed': 1
}

train_data.Status.replace(Status_mapping,inplace=True)
test_data.Status.replace(Status_mapping,inplace=True)


train_data



Country  Year  Status  Population  Hepatitis B  Measles  \
Country                                                                     
Afghanistan 0           0  2015       0  33736494.0         65.0     1154   
            1           0  2014       0    327582.0         62.0      492   
            2           0  2013       0  31731688.0         64.0      430   
            3           0  2012       0   3696958.0         67.0     2787   
            4           0  2011       0   2978599.0         68.0     3013   
...                   ...   ...     ...         ...          ...      ...   
Zimbabwe    2843      177  2004       0  12777511.0         68.0       31   
            2844      177  2003       0  12633897.0          7.0      998   
            2845      177  2002       0    125525.0         73.0      304   
            2846      177  2001       0  12366165.0         76.0      529   
            2847      177  2000       0  12222251.0         79.0     1483   

                  Polio  Diphtheria  HIV/AIDS  infant deaths  \
Country                                                        
Afghanistan 0       6.0        65.0       0.1             62   
            1      58.0        62.0       0.1             64   
            2      62.0        64.0       0.1             66   
            3      67.0        67.0       0.1             69   
            4      68.0        68.0       0.1             71   
...                 ...         ...       ...            ...   
Zimbabwe    2843   67.0        65.0      33.6             27   
            2844    7.0        68.0      36.7             26   
            2845   73.0        71.0      39.8             25   
            2846   76.0        75.0      42.1             25   
            2847   78.0        78.0      43.5             24   

                  under-five deaths  Total expenditure         GDP   BMI  \
Country                                                                    
Afghanistan 0                    83               8.16  584.259210  19.1   
            1                    86               8.18  612.696514  18.6   
            2                    89               8.13  631.744976  18.1   
            3                    93               8.52  669.959000  17.6   
            4                    97               7.87   63.537231  17.2   
...                             ...                ...         ...   ...   
Zimbabwe    2843                 42               7.13  454.366654  27.1   
            2844                 41               6.52  453.351155  26.7   
            2845                 40               6.53   57.348340  26.3   
            2846                 39               6.16  548.587312  25.9   
            2847                 39               7.10  547.358878  25.5   

                  thinness  1-19 years  Alcohol  Schooling  Life expectancy  
Country                                                                      
Afghanistan 0                     17.2     0.01       10.1             65.0  
            1                     17.5     0.01       10.0             59.9  
            2                     17.7     0.01        9.9             59.9  
            3                     17.9     0.01        9.8             59.5  
            4                     18.2     0.01        9.5             59.2  
...                                ...      ...        ...              ...  
Zimbabwe    2843                   9.4     4.36        9.2             44.3  
            2844                   9.8     4.06        9.5             44.5  
            2845                   1.2     4.43       10.0             44.8  
            2846                   1.6     1.72        9.8             45.3  
            2847                  11.0     1.68        9.8             46.0  

[2848 rows x 18 columns]

In [120]:
train_data.info()


<class 'pandas.core.frame.DataFrame'>
MultiIndex: 2848 entries, ('Afghanistan', 0) to ('Zimbabwe', 2847)
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Country               2848 non-null   int64  
 1   Year                  2848 non-null   int64  
 2   Status                2848 non-null   int64  
 3   Population            2848 non-null   float64
 4   Hepatitis B           2848 non-null   float64
 5   Measles               2848 non-null   int64  
 6   Polio                 2848 non-null   float64
 7   Diphtheria            2848 non-null   float64
 8   HIV/AIDS              2848 non-null   float64
 9   infant deaths         2848 non-null   int64  
 10  under-five deaths     2848 non-null   int64  
 11  Total expenditure     2848 non-null   float64
 12  GDP                   2848 non-null   float64
 13  BMI                   2848 non-null   float64
 14  thinness  1-19 years  2848 non-null   flo

In [121]:
train_data.reset_index(drop=True, inplace=True)
train_data.sort_index(inplace=True,ascending=True)

test_data.reset_index(drop=True, inplace=True)
test_data.sort_index(inplace=True,ascending=True)

In [122]:
train_data.drop(['Country'], axis=1,inplace=True)
test_data.drop(['Country'], axis=1,inplace=True)

In [123]:
from sklearn.model_selection import train_test_split

x_train, x_val, y_train, y_val = train_test_split(train_data.drop(['Life expectancy'], axis=1), train_data['Life expectancy'], test_size=.1, random_state=42)


In [124]:

from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
X_train = std_scaler.fit_transform(x_train)
X_val = std_scaler.transform(x_val)

X_test = std_scaler.transform(test_data)

In [125]:
X_train
X_val

array([[ 0.18609453, -0.10512537,  2.17278212, ..., -0.92913299,
         0.68636352,  0.71952354],
       [-1.33133832,  0.76542816, -0.46023943, ..., -0.50088617,
        -0.00665787,  0.40929748],
       [ 0.84753961, -0.75804051, -0.46023943, ..., -1.04182952,
        -0.25821723,  0.09907141],
       ...,
       [-0.45589629, -0.10512537, -0.46023943, ...,  0.8740115 ,
        -0.4160584 , -1.01774241],
       [-0.45589629,  0.76542816, -0.46023943, ...,  0.67115775,
        -0.27054857, -0.9556972 ],
       [-0.65043896, -0.32276375, -0.46023943, ..., -0.95167229,
         1.90716631,  1.24690785]])

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
مدل‌سازی
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    حال که داده را پاکسازی کرده و احتمالا ویژگی‌هایی را به آن افزوده یا از آن حذف کرده‌اید، وقت آن است که مدلی آموزش دهید که بتواند متغیر هدف این مسئله را پیش‌بینی کند.
</font>
</p>

<h3 align=right style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    استفاده از <code>scikit-learn</code>
</font>
</h3>


<div dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    برای الگوریتم رگرسیون چند‌جمله‌ای پکیج یکتایی در کتابخانه‌ی <code>scikit-learn</code> وجود ندارد. بنابراین برای استفاده از این الگوریتم، می‌بایست ابتدا ویژگی‌های دیتافریم را با استفاده از کلاس <code>PolynomialFeatures</code> که در پکیج <code>preprocessing</code> قرار دارد، به ویژگی‌های چندجمله‌ای تبدیل کنید. سپس از آن ویژگی‌های تغییریافته (<code>transformed</code>) برای آموزش یک رگرسیون خطی معمولی استفاده کنید.
    <br>
    برخی از آرگومان‌های مهم این کلاس در جدول زیر آمده است، اما برای مطالعه کامل‌تر مستندات می‌توانید به <a href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html">این لینک</a> مراجعه فرمایید.

</font>
</div>
    
<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
<br>

|آرگومان|نوع|توضیحات|
|:------:|:--------:|:---:|
|<code>degree</code>|<code>int</code> یا <code>tuple</code>|درجه چند جمله‌ای را نشان می‌دهد. به‌صورت پیش‌فرض عدد <code>2</code> است.<br> اگر به شکل <code>tuple</code> باشد، به شکل <code dir=ltr>(min_degree, max_degree)</code> تفسیر می‌شود.|
|<code>include_bias</code>|<code>bool</code>|آیا شامل بایاس باشد یا خیر. <br> در صورت <code>True</code> بودن، عدد <code>1</code> هم به چندجمله‌ای افزوده می‌شود. <br> (به این معنی که چندجمله‌ای‌های تولید‌شده حداقل از درجه‌ی <code>min_degree</code> و حداکثر از مرتبه‌ی <code>max_degree</code> خواهند بود.)|
    
</font>
</div>
</center>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    کلاس <code>PolynomialFeatures</code> دارای سه متد به شرح زیر است: 
</font>
</p>


<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|اسم متد|توضیحات|
|:------:|:---:|
|<code>fit</code>|تعداد جملات چندجمله‌ای خروجی را حساب می‌کند|
|<code>transform</code>|داده‌ی ورودی را به جملات چند‌جمله‌ای تبدیل می‌کند|
|<code>fit_transform</code>|ابتدا تابع <code>fit</code> و سپس تابع <code>transform</code> را اجرا می‌کند|

</font>
</div>
</center>

<h3 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
آموزش مدل
</font>
</h3>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    در سلول پایین با استفاده از کتابخانه‌ی <code>scikit-learn</code> مدلی آموزش دهید که بتواند متغیر هدف این مسئله را پیش‌بینی کند.
    <br>
    برای ساختن ویژگی‌های چندجمله‌ای، شیء <code>poly_transformer</code> را مقداردهی کنید. بعداً در سلول جواب‌ساز این شیء ذخیره شده و در زمان داوری، مورد ارزیابی واقع خواهد شد.
    <br>
     سپس با ویژگی‌های چندجمله‌ای خروجی <code>poly_transformer</code> مدل رگرسیون خطی را آموزش دهید و آن را در شیء <code>model</code> ذخیره کنید. متغیر <code>model</code> در فرآیند داوری مورد ارزیابی واقع خواهد شد. بنابراین اطمینان حاصل کنید که از جنس <code>LinearRegression</code> باشد؛ در غیر این صورت نمره‌ای دریافت نخواهید کرد!
</font>
</p>

In [126]:
from sklearn.preprocessing import PolynomialFeatures

poly_transformer = PolynomialFeatures(degree=2, include_bias=True)
x_train_trans = poly_transformer.fit_transform(X_train)
x_val_trans = poly_transformer.transform(X_val)

In [127]:
from sklearn.linear_model import LinearRegression



model = LinearRegression()
model.fit(x_train_trans, y_train)
y_pred = model.predict(x_val_trans)


<h3 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
معیار ارزیابی
</font>
</h3>
<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    معیاری که برای ارزیابی عملکرد مدل انتخاب کرده‌ایم، <code>r2_score</code> نام دارد.
    <br>
    این معیار، سنجه ارزیابی کیفیت مدل شماست. به عبارت بهتر در سامانه داوری هم از همین معیار برای نمره‌دهی استفاده شده است.
    <br>
    پیشنهاد می‌شود با توجه به این معیار، عملکرد مدل خود را بر روی مجموعه‌ی آموزش یا اعتبارسنجی ارزیابی کنید.
</font>
</p>

In [128]:
# evaluate your model
from sklearn.metrics import r2_score
r2_score(y_val, y_pred)
# To-Do



0.8933577494025479

In [129]:
# for test data

x_test_trans = poly_transformer.transform(X_test)




model = LinearRegression()
model.fit(x_train_trans, y_train)
y_pred = model.predict(x_test_trans)
y_pred.round(2)


array([66.43, 65.87, 65.24, 64.6 , 63.92, 62.46, 61.63, 59.21, 58.47,
       53.55, 52.71, 52.38, 51.43, 50.81, 50.41, 51.89, 63.44, 58.25,
       68.5 , 61.62, 57.91, 63.68, 54.97, 54.35, 54.69, 55.26, 48.68,
       50.91, 45.64, 39.77, 41.85, 39.17, 60.89, 62.06, 59.54, 57.59,
       54.29, 53.75, 58.58, 58.69, 56.62, 55.32, 56.99, 58.05, 58.15,
       56.95, 60.92, 61.98, 81.21, 80.28, 80.1 , 80.32, 77.98, 78.12,
       77.98, 76.69, 75.63, 76.78, 76.71, 76.8 , 76.41, 75.93, 75.56,
       74.87, 67.86, 64.69, 65.06, 64.65, 64.57, 63.62, 64.07, 62.67,
       62.42, 61.46, 60.73, 53.88, 59.66, 55.72, 58.  , 59.26])

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font color="red"><b color='red'>توجه:</b></font>
<font face="vazir" size=3>
 جهت کسب امتیاز کامل نیاز است تا پاسخ شما حداقل امتیاز <code>50</code> را با توجه به معیار معرفی‌شده کسب نماید.
</font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
 پیش‌بینی برای داده تست و خروجی
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    پس از مهندسی ویژگی و مدلسازی، الگوریتمی دارید که می‌تواند شما را از متغیرهای مستقل به متغیر هدف برساند.
    <br>
    از این مدل برای پیش‌بینی نمونه‌های موجود در مجموعه‌ی آزمون استفاده کنید و نتایج را در قالب جدول (<code>dataframe</code>) زیر آماده کنید.
</font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>Life expectancy</code>|امید به زندگی پیش‌بینی‌شده|
    
</font>
</div>
</center>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    توجه داشته باشید که نام دیتافریم باید <code>submission</code> باشد؛ در غیر این‌صورت، سامانه‌ی داوری قادر به ارزیابی خروجی شما نخواهد بود.
    این دیتافریم تنها شامل ۱ ستون با اسم <code>Life expectancy</code> است و ۸۰ سطر دارد.
    <br>
    به ازای هر سطر موجود در دیتافریم <code>test</code>، باید یک مقدار پیش‌بینی‌شده داشته باشید.
    به‌عنوان مثال جدول زیر، ۵ سطر ابتدایی دیتافریم <code>submission</code> را نشان می‌دهد. البته در جواب شما، اعداد ستون <code>Life expectancy</code> ممکن است متفاوت باشد.
</font>
</p>

<center>
<div style="direction: ltr;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
||<code>Life expectancy</code>|
|:----:|:-----:|
|0|70.1|
|1|65.2|
|2|50.0|
|3|81.72|
|4|75.25|

</font>
</div>
</center>

In [130]:

submission = pd.DataFrame()
submission['Life expectancy'] = y_pred.round(2)
submission

Life expectancy
0             66.43
1             65.87
2             65.24
3             64.60
4             63.92
..              ...
75            53.88
76            59.66
77            55.72
78            58.00
79            59.26

[80 rows x 1 columns]

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
<b>سلول جواب‌ساز</b>
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    برای ساخته‌شدن فایل <code>result.zip</code> سلول زیر را اجرا کنید. توجه داشته باشید که پیش از اجرای سلول زیر تغییرات اعمال شده در نت‌بوک را ذخیره کرده باشید (<code>ctrl+s</code>) تا در صورت نیاز به پشتیبانی امکان بررسی کد شما وجود داشته باشد.
</font>
</p>

In [131]:
import zipfile
import joblib

def compress(file_names):
    print("File Paths:")
    print(file_names)
    compression = zipfile.ZIP_DEFLATED
    with zipfile.ZipFile("result.zip", mode="w") as zf:
        for file_name in file_names:
            zf.write('./' + file_name, file_name, compress_type=compression)


joblib.dump(poly_transformer, 'poly_transformer')
joblib.dump(model, 'model')
submission.to_csv('submission.csv', index=False)

file_names = ['poly_transformer', 'model', 'submission.csv', 'life_expectancy.ipynb']
compress(file_names)

File Paths:
['poly_transformer', 'model', 'submission.csv', 'life_expectancy.ipynb']


<h4 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
راهنمایی
</font>
</h4>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    ۱. در مراحل پیش‌پردازش از تغییر مقیاس ویژگی استفاده کنید.
    <br>
    ۲. مقادیر گم‌شده‌ی ویژگی‌ها را بررسی و مدیریت کنید.
</font>
</p>